openapi.yaml

```diff
@@ -0,0 +1,2411 @@
+      operationId: getDagRuns
+      tags: [DAGRun]
+      parameters:
+        - $ref: '#/components/parameters/PageLimit'
+        - $ref: '#/components/parameters/PageOffset'
+        - $ref: '#/components/parameters/FilterExecutionDateGTE'
+        - $ref: '#/components/parameters/FilterExecutionDateLTE'
+        - $ref: '#/components/parameters/FilterStartDateGTE'
+        - $ref: '#/components/parameters/FilterStartDateLTE'
+        - $ref: '#/components/parameters/FilterEndDateGTE'
+        - $ref: '#/components/parameters/FilterEndDateLTE'
+      responses:
+        '200':
+          description: List of DAG Runs.
+          content:
+            application/json:
+              schema:
+                allOf:
+                  - $ref: '#/components/schemas/DAGRunCollection'
+                  - $ref: '#/components/schemas/CollectionInfo'
+        '401':
+          $ref: '#/components/responses/Unauthenticated'
+
+  /dags/~/dagRuns/list:
+    post:
+      summary: Get all DAG Runs.
+      description: >
+        It allows you to fetch DAG Runs from many DAGs.
+      operationId: getDagRunsBatch
+      tags: [DAGRun]
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/ListDagRunsForm'
+          application/x-www-form-urlencoded:
+            schema:
+              $ref: '#/components/schemas/ListDagRunsForm'
+      responses:
+        '200':
+          description: List of DAG Runs.
+          content:
+            application/json:
+              schema:
+                allOf:
+                  - $ref: '#/components/schemas/DAGRunCollection'
+                  - $ref: '#/components/schemas/CollectionInfo'
+        '400':
+          $ref: '#/components/responses/BadRequest'
+        '401':
+          $ref: '#/components/responses/Unauthenticated'
+        '403':
+          $ref: '#/components/responses/PermissionDenied'
+
+  /dags/{dag_id}/dagRuns/{execution_date}:
+    parameters:
+      - $ref: '#/components/parameters/DAGID'
+      - $ref: '#/components/parameters/ExecutionDate'
+
+    get:
+      summary: Get a DAG Run
+      operationId: getDagRun
+      tags: [DAGRun]
+      responses:
+        '200':
+          description: Successful response.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/DAGRun'
+        '401':
+          $ref: '#/components/responses/Unauthenticated'
+        '403':
+          $ref: '#/components/responses/PermissionDenied'
+        '404':
+          $ref: '#/components/responses/NotFound'
+
+    post:
+      summary: Trigger a DAG Run
+      operationId: createDagRun
+      tags: [DAGRun]
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/DAGRun'
```

Comment:
Same comment here about conf parameter (accepting a JSON `string` vs accepting a `type: object` and then doing a json.dumps to store it)

Reply:
This is needed for strongly typed languages. If we want the API to be usable by every client, we must provide a scheme for each field. If we are unable to provide schema then we must return JSON as a string.

Reply:
The schema can be `object`? We don't have to specify more than that.

It is totally possible to encode arbitrary data in staticly typed languages https://blog.golang.org/json#TOC_5. for example.

Reply:
I'm afraid that it can still cause various problems. When I write about strongly typed language, I mean Java/C++. In Java,  there are two different libraries that are written with a completely different philosophy.
* Unstructured JSON e.g. Json-Simple API - This creates special JSON objects that are used to store data. Thanks to this you can process any objects, but the API is not very pleasant and you don't have type hints. This is similar to the approach available in Python.
```
import java.io.FileReader; 
import java.util.Iterator; 
import java.util.Map; 
  
import org.json.simple.JSONArray; 
import org.json.simple.JSONObject; 
import org.json.simple.parser.*; 
  
public class JSONReadExample  
{ 
    public static void main(String[] args) throws Exception  
    { 
        // parsing file "JSONExample.json" 
        Object obj = new JSONParser().parse(new FileReader("JSONExample.json")); 
          
        // typecasting obj to JSONObject 
        JSONObject jo = (JSONObject) obj; 
          
        // getting firstName and lastName 
        String firstName = (String) jo.get("firstName"); 
        String lastName = (String) jo.get("lastName"); 
          
        System.out.println(firstName); 
        System.out.println(lastName); 
          
        // getting age 
        long age = (long) jo.get("age"); 
        System.out.println(age); 
          
        // getting address 
        Map address = ((Map)jo.get("address")); 
          
        // iterating address Map 
        Iterator<Map.Entry> itr1 = address.entrySet().iterator(); 
        while (itr1.hasNext()) { 
            Map.Entry pair = itr1.next(); 
            System.out.println(pair.getKey() + " : " + pair.getValue()); 
        } 
          
        // getting phoneNumbers 
        JSONArray ja = (JSONArray) jo.get("phoneNumbers"); 
          
        // iterating phoneNumbers 
        Iterator itr2 = ja.iterator(); 
          
        while (itr2.hasNext())  
        { 
            itr1 = ((Map) itr2.next()).entrySet().iterator(); 
            while (itr1.hasNext()) { 
                Map.Entry pair = itr1.next(); 
                System.out.println(pair.getKey() + " : " + pair.getValue()); 
            } 
        } 
    } 
} 
```
* Structured JSON e.g. GSON -  An approach that requires you to define classes for all objects before using them. This means that you must convert OpneAPI schemas to the corresponding class languages before use.
```
public class Car {
    public String brand = null;
    public int    doors = 0;
}
String json = "{\"brand\":\"Jeep\", \"doors\": 3}";

Gson gson = new Gson();

Car car = gson.fromJson(json, Car.class);
```
This approach is very popular in the Java world because it provides type hints and is closest to the philosophy of the language.

In our API it is necessary to use both approaches because for conf field we are not able to specify a specific type. The easiest way to do this is to use two different libraries.

Of course, there are workarounds that can also process unstructured JSON in a GSON also, but we cannot guarantee that such a workaround will exist in every language and library.  The processing of structured JSON objects and unstructured JSON objects are completely two different problems, and most often completely different libraries are used.

Reply:
I tested it using generated apis and seems that it is super-easy to get it in Java using Jackson (this is what generated API uses).

Seems that both read and write is simple and straightforward (and @houqp also wrote a Go example showing that it's also easy and straightforward). 

In this case I am for "object" 

```
 @Test
    public void dagTest() throws IOException {
        DAGRun dagRun = new DAGRun();
        HashMap<Object, Object> confObject = new HashMap<>();
        String array[] = new String[] {"1", "2", "3"};
        confObject.put("a", array);
        confObject.put("b", "c");
        dagRun.setConf(confObject);
        ObjectMapper mapper = api.getApiClient().getObjectMapper();
        String data = mapper.writer().writeValueAsString(dagRun);
        System.out.println(data);
        TypeReference<DAGRun> typeRef = new TypeReference<DAGRun>() {};
        DAGRun dagRunRead = mapper.readValue(data, typeRef);
        System.out.println();
        System.out.println(dagRunRead);
        System.out.println(dagRunRead.getConf().getClass());

        DAGRun dagRun2 = new DAGRun();
        HashMap confObject2 =
                mapper.readValue("{\"a\":[\"1\", \"2\", \"3\"],\"b\":\"c\" }", confObject.getClass());
        dagRun2.setConf(confObject2);
        String data2 = mapper.writer().writeValueAsString(dagRun2);
        System.out.println(data2);
        DAGRun dagRunRead2 = mapper.readValue(data, typeRef);
        System.out.println();
        System.out.println(dagRunRead2);
        System.out.println(dagRunRead2.getConf().getClass());
}
```

Produces:

```
"dag_run_id":null,"dag_id":null,"execution_date":null,"start_date":null,"end_date":null,"state":null,"external_trigger":true,"conf":{"a":["1","2","3"],"b":"c"}}

class DAGRun {
    dagRunId: null
    dagId: null
    executionDate: null
    startDate: null
    endDate: null
    state: null
    externalTrigger: true
    conf: {a=[1, 2, 3], b=c}
}
class java.util.LinkedHashMap
{"dag_run_id":null,"dag_id":null,"execution_date":null,"start_date":null,"end_date":null,"state":null,"external_trigger":true,"conf":{"a":["1","2","3"],"b":"c"}}

class DAGRun {
    dagRunId: null
    dagId: null
    executionDate: null
    startDate: null
    endDate: null
    state: null
    externalTrigger: true
    conf: {a=[1, 2, 3], b=c}
}
class java.util.LinkedHashMap
```